File: app/src/main/java/com/example/nasmovie/view/BottomSheetDrawer.java
package com.example.nasmovie.view;

import android.animation.ValueAnimator;
import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.nasmovie.R;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.ArrayList;
import java.util.List;

/**
 * 通用底部抽屉组件
 * 支持展开状态下继续向上拖动时的拉伸缩放效果
 */
public class BottomSheetDrawer extends BottomSheetDialogFragment {

    private static final String TAG = "BottomSheetDrawer";

    private final List<MenuItem> items = new ArrayList<>();
    private View contentView;
    private View bottomSheetView;

    // 拉伸效果相关
    private static final float SCALE_RATIO = 0.4f;      // 拉伸系数
    private static final float MAX_SCALE = 1.15f;       // 最大缩放倍数
    private static final float REPLY_RATIO = 0.3f;      // 回弹时间系数

    private float lastTouchY;
    private float stretchStartY;  // 开始拉伸时的 Y 坐标
    private boolean isStretching = false;
    private float currentScale = 1.0f;
    private float originalHeight;
    private float originalBottomSheetHeight;  // bottomSheet 原始高度
    private int initialTopPosition = -1;  // bottomSheet 初始顶部位置

    public static BottomSheetDrawer newInstance() {
        return new BottomSheetDrawer();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(STYLE_NORMAL, R.style.Theme_NASMovie_BottomSheetDialog);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        // 创建自定义 BottomSheetDialog 以拦截触摸事件
        return new StretchBottomSheetDialog(requireContext(), getTheme());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        contentView = inflater.inflate(R.layout.sheet_bottom_drawer, container, false);
        return contentView;
    }
    
    /**
     * 自定义 BottomSheetDialog，重写 dispatchTouchEvent 来拦截触摸事件
     */
    private class StretchBottomSheetDialog extends BottomSheetDialog {
        
        public StretchBottomSheetDialog(@NonNull Context context, int theme) {
            super(context, theme);
        }
        
        @Override
        public boolean dispatchTouchEvent(@NonNull MotionEvent ev) {
            // 先让拉伸处理逻辑处理
            if (handleStretchTouch(ev)) {
                return true;  // 事件被消费
            }
            return super.dispatchTouchEvent(ev);
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        
        // 获取 BottomSheet 的根视图
        View parent = (View) contentView.getParent();
        if (parent != null) {
            bottomSheetView = parent;
            
            // 设置不裁剪子View，让拉伸内容可以超出边界显示
            if (bottomSheetView instanceof ViewGroup) {
                ((ViewGroup) bottomSheetView).setClipChildren(false);
            }
            // LinearLayout 也是 ViewGroup
            if (contentView instanceof ViewGroup) {
                ((ViewGroup) contentView).setClipChildren(false);
            }
            if (bottomSheetView.getParent() instanceof ViewGroup) {
                ViewGroup grandParent = (ViewGroup) bottomSheetView.getParent();
                grandParent.setClipChildren(false);
                grandParent.setClipToPadding(false);
            }
            
            // 记录原始尺寸
            bottomSheetView.post(() -> {
                originalHeight = bottomSheetView.getHeight();
                originalBottomSheetHeight = bottomSheetView.getHeight();
            });
        }
        
        // 设置 skipCollapsed 并强制展开
        BottomSheetDialog dialog = (BottomSheetDialog) getDialog();
        if (dialog != null) {
            dialog.getBehavior().setSkipCollapsed(true);
            dialog.getBehavior().setState(BottomSheetBehavior.STATE_EXPANDED);
            
            // 在平板设备上设置宽度为屏幕宽度
            if (bottomSheetView != null) {
                android.util.DisplayMetrics metrics = new android.util.DisplayMetrics();
                dialog.getWindow().getWindowManager().getDefaultDisplay().getMetrics(metrics);
                int screenWidth = metrics.widthPixels;
                bottomSheetView.getLayoutParams().width = screenWidth;
            }
        }
    }

    private boolean handleStretchTouch(MotionEvent event) {
        // 获取 bottomSheet 当前位置
        int[] location = new int[2];
        if (bottomSheetView != null) {
            bottomSheetView.getLocationOnScreen(location);
        }
        int topPosition = location[1];
        
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                lastTouchY = event.getRawY();
                initialTopPosition = topPosition;
                return false;

            case MotionEvent.ACTION_MOVE:
                float deltaY = lastTouchY - event.getRawY();  // 向上为正
                
                if (isStretching) {
                    // 已经在拉伸状态，继续处理拉伸
                    float totalDeltaY = stretchStartY - event.getRawY();  // 向上为正
                    if (totalDeltaY > 0) {
                        float stretchDistance = totalDeltaY * SCALE_RATIO;
                        applyScale(stretchDistance);
                    } else if (totalDeltaY < -10) {
                        // 向下拖动超过阈值，取消拉伸状态
                        contentView.setScaleY(1.0f);
                        currentScale = 1.0f;
                        isStretching = false;
                    } else {
                        contentView.setScaleY(1.0f);
                        currentScale = 1.0f;
                    }
                    lastTouchY = event.getRawY();
                    return true;
                }
                
                // 检测是否可以触发拉伸：
                // 用户向上拖动，但 topPosition 没有变小（说明抽屉已经无法继续向上移动）
                if (deltaY > 10 && topPosition >= initialTopPosition) {
                    isStretching = true;
                    stretchStartY = lastTouchY;
                    lastTouchY = event.getRawY();
                    return true;
                }
                
                // 正常拖动：更新位置
                initialTopPosition = topPosition;
                lastTouchY = event.getRawY();
                return false;

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                if (isStretching) {
                    isStretching = false;
                    animateRebound();
                    return true;
                }
                break;
        }
        return false;
    }

    /**
     * 重置 bottomSheet 位置
     */
    private void resetBottomSheetHeight() {
        // 不需要重置，因为不再移动 bottomSheet
    }

    /**
     * 应用缩放效果
     */
    private void applyScale(float stretchDistance) {
        float newScale = 1.0f + (stretchDistance / originalHeight);

        // 限制最大缩放倍数
        newScale = Math.min(newScale, MAX_SCALE);
        newScale = Math.max(newScale, 1.0f);

        // 设置缩放锚点为底部，使缩放时内容向上延伸
        contentView.setPivotY(contentView.getHeight());
        
        currentScale = newScale;
        contentView.setScaleY(newScale);
        
        // 不再移动 bottomSheet，依靠 clipChildren=false 让内容超出边界显示
    }

    /**
     * 回弹动画
     */
    private void animateRebound() {
        float startScale = currentScale;
        ValueAnimator animator = ValueAnimator.ofFloat(startScale, 1.0f);
        animator.setDuration((long) ((startScale - 1.0f) * 500 / REPLY_RATIO));
        animator.addUpdateListener(animation -> {
            float scale = (float) animation.getAnimatedValue();
            contentView.setScaleY(scale);
            currentScale = scale;
        });
        animator.start();
        currentScale = 1.0f;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        LinearLayout container = view.findViewById(R.id.itemsContainer);

        for (int i = 0; i < items.size(); i++) {
            MenuItem item = items.get(i);
            View itemView = LayoutInflater.from(getContext())
                    .inflate(R.layout.item_bottom_drawer, container, false);

            TextView textView = itemView.findViewById(R.id.itemText);
            textView.setText(item.title);

            if (item.isDestructive) {
                textView.setTextColor(Color.parseColor("#ff716c"));
            }

            itemView.setOnClickListener(v -> {
                dismiss();
                if (item.clickListener != null) {
                    item.clickListener.onClick();
                }
            });

            container.addView(itemView);
        }
    }

    public BottomSheetDrawer addItem(String title, OnItemClickListener listener) {
        items.add(new MenuItem(title, false, listener));
        return this;
    }

    public BottomSheetDrawer addDestructiveItem(String title, OnItemClickListener listener) {
        items.add(new MenuItem(title, true, listener));
        return this;
    }

    public interface OnItemClickListener {
        void onClick();
    }

    private static class MenuItem {
        final String title;
        final boolean isDestructive;
        final OnItemClickListener clickListener;

        MenuItem(String title, boolean isDestructive, OnItemClickListener clickListener) {
            this.title = title;
            this.isDestructive = isDestructive;
            this.clickListener = clickListener;
        }
    }

    public static class Builder {
        private final List<MenuItem> items = new ArrayList<>();

        public Builder addItem(String title, OnItemClickListener listener) {
            items.add(new MenuItem(title, false, listener));
            return this;
        }

        public Builder addDestructiveItem(String title, OnItemClickListener listener) {
            items.add(new MenuItem(title, true, listener));
            return this;
        }

        public BottomSheetDrawer build() {
            BottomSheetDrawer drawer = new BottomSheetDrawer();
            drawer.items.addAll(this.items);
            return drawer;
        }
    }
}

File: app/src/main/java/com/example/nasmovie/view/BottomSheetDrawer.kt
package com.example.nasmovie.view

import android.animation.ValueAnimator
import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.example.nasmovie.R
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment

/**
 * 通用底部抽屉组件
 * 支持展开状态下继续向上拖动时的拉伸缩放效果
 */
class BottomSheetDrawer : BottomSheetDialogFragment() {

    companion object {
        private const val TAG = "BottomSheetDrawer"

        fun newInstance(): BottomSheetDrawer = BottomSheetDrawer()
    }

    private val items = mutableListOf<MenuItem>()
    private var contentView: View? = null
    private var bottomSheetView: View? = null

    // 拉伸效果相关
    private val scaleRatio = 0.4f       // 拉伸系数
    private val maxScale = 1.15f        // 最大缩放倍数
    private val replyRatio = 0.3f       // 回弹时间系数

    private var lastTouchY = 0f
    private var stretchStartY = 0f      // 开始拉伸时的 Y 坐标
    private var isStretching = false
    private var currentScale = 1.0f
    private var originalHeight = 0f
    private var originalBottomSheetHeight = 0f   // bottomSheet 原始高度
    private var initialTopPosition = -1          // bottomSheet 初始顶部位置

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NORMAL, R.style.Theme_NASMovie_BottomSheetDialog)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        // 创建自定义 BottomSheetDialog 以拦截触摸事件
        return StretchBottomSheetDialog(requireContext(), theme)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        contentView = inflater.inflate(R.layout.sheet_bottom_drawer, container, false)
        return contentView
    }

    /**
     * 自定义 BottomSheetDialog，重写 dispatchTouchEvent 来拦截触摸事件
     */
    private inner class StretchBottomSheetDialog(
        context: Context,
        theme: Int
    ) : BottomSheetDialog(context, theme) {

        override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
            // 先让拉伸处理逻辑处理
            if (handleStretchTouch(ev)) {
                return true  // 事件被消费
            }
            return super.dispatchTouchEvent(ev)
        }
    }

    override fun onStart() {
        super.onStart()

        // 获取 BottomSheet 的根视图
        val parent = contentView?.parent as? View
        if (parent != null) {
            bottomSheetView = parent

            // 设置不裁剪子View，让拉伸内容可以超出边界显示
            (bottomSheetView as? ViewGroup)?.clipChildren = false
            (contentView as? ViewGroup)?.clipChildren = false
            (bottomSheetView?.parent as? ViewGroup)?.let { grandParent ->
                grandParent.clipChildren = false
                grandParent.clipToPadding = false
            }

            // 记录原始尺寸
            bottomSheetView?.post {
                bottomSheetView?.let {
                    originalHeight = it.height.toFloat()
                    originalBottomSheetHeight = it.height.toFloat()
                }
            }
        }

        // 设置 skipCollapsed 并强制展开
        (dialog as? BottomSheetDialog)?.let { dialog ->
            dialog.behavior.skipCollapsed = true
            dialog.behavior.state = BottomSheetBehavior.STATE_EXPANDED

            // 在平板设备上设置宽度为屏幕宽度
            bottomSheetView?.let { sheet ->
                sheet.layoutParams.width = getScreenWidth()
            }
        }
    }

    /**
     * 获取屏幕宽度，兼容不同 Android 版本
     */
    @Suppress("DEPRECATION")
    private fun getScreenWidth(): Int {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // Android 11+ 使用 WindowMetrics
            val windowMetrics = requireContext()
                .getSystemService(android.view.WindowManager::class.java)
                .currentWindowMetrics
            windowMetrics.bounds.width()
        } else {
            // Android 10 及以下使用旧 API
            val metrics = android.util.DisplayMetrics()
            val windowManager = requireContext()
                .getSystemService(Context.WINDOW_SERVICE) as? android.view.WindowManager
            windowManager?.defaultDisplay?.getMetrics(metrics)
            metrics.widthPixels
        }
    }

    private fun handleStretchTouch(event: MotionEvent): Boolean {
        // 获取 bottomSheet 当前位置
        val location = IntArray(2)
        bottomSheetView?.getLocationOnScreen(location)
        val topPosition = location[1]

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                lastTouchY = event.rawY
                initialTopPosition = topPosition
                return false
            }

            MotionEvent.ACTION_MOVE -> {
                val deltaY = lastTouchY - event.rawY  // 向上为正

                if (isStretching) {
                    // 已经在拉伸状态，继续处理拉伸
                    val totalDeltaY = stretchStartY - event.rawY  // 向上为正
                    if (totalDeltaY > 0) {
                        val stretchDistance = totalDeltaY * scaleRatio
                        applyScale(stretchDistance)
                    } else if (totalDeltaY < -10) {
                        // 向下拖动超过阈值，取消拉伸状态
                        contentView?.scaleY = 1.0f
                        currentScale = 1.0f
                        isStretching = false
                    } else {
                        contentView?.scaleY = 1.0f
                        currentScale = 1.0f
                    }
                    lastTouchY = event.rawY
                    return true
                }

                // 检测是否可以触发拉伸：
                // 用户向上拖动，但 topPosition 没有变小（说明抽屉已经无法继续向上移动）
                if (deltaY > 10 && topPosition >= initialTopPosition) {
                    isStretching = true
                    stretchStartY = lastTouchY
                    lastTouchY = event.rawY
                    return true
                }

                // 正常拖动：更新位置
                initialTopPosition = topPosition
                lastTouchY = event.rawY
                return false
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                if (isStretching) {
                    isStretching = false
                    animateRebound()
                    return true
                }
            }
        }
        return false
    }

    /**
     * 应用缩放效果
     */
    private fun applyScale(stretchDistance: Float) {
        var newScale = 1.0f + (stretchDistance / originalHeight)

        // 限制最大缩放倍数
        newScale = newScale.coerceAtMost(maxScale)
        newScale = newScale.coerceAtLeast(1.0f)

        // 设置缩放锚点为底部，使缩放时内容向上延伸
        contentView?.let { view ->
            view.pivotY = view.height.toFloat()
            currentScale = newScale
            view.scaleY = newScale
        }
    }

    /**
     * 回弹动画
     */
    private fun animateRebound() {
        val startScale = currentScale
        val animator = ValueAnimator.ofFloat(startScale, 1.0f)
        animator.duration = ((startScale - 1.0f) * 500 / replyRatio).toLong()
        animator.addUpdateListener { animation ->
            val scale = animation.animatedValue as Float
            contentView?.scaleY = scale
            currentScale = scale
        }
        animator.start()
        currentScale = 1.0f
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val container = view.findViewById<LinearLayout>(R.id.itemsContainer)

        for (item in items) {
            val itemView = LayoutInflater.from(context)
                .inflate(R.layout.item_bottom_drawer, container, false)

            val textView = itemView.findViewById<TextView>(R.id.itemText)
            textView.text = item.title

            if (item.isDestructive) {
                textView.setTextColor(Color.parseColor("#ff716c"))
            }

            itemView.setOnClickListener {
                dismiss()
                item.clickListener?.invoke()
            }

            container.addView(itemView)
        }
    }

    fun addItem(title: String, listener: (() -> Unit)?): BottomSheetDrawer {
        items.add(MenuItem(title, false, listener))
        return this
    }

    fun addDestructiveItem(title: String, listener: (() -> Unit)?): BottomSheetDrawer {
        items.add(MenuItem(title, true, listener))
        return this
    }

    fun interface OnItemClickListener {
        fun onClick()
    }

    private data class MenuItem(
        val title: String,
        val isDestructive: Boolean,
        val clickListener: (() -> Unit)?
    )

    class Builder {
        private val items = mutableListOf<MenuItem>()

        fun addItem(title: String, listener: (() -> Unit)?): Builder {
            items.add(MenuItem(title, false, listener))
            return this
        }

        fun addDestructiveItem(title: String, listener: (() -> Unit)?): Builder {
            items.add(MenuItem(title, true, listener))
            return this
        }

        fun build(): BottomSheetDrawer {
            return BottomSheetDrawer().apply {
                this.items.addAll(this@Builder.items)
            }
        }
    }
}